.github/workflows/cabal.yml

```diff
@@ -29,32 +29,41 @@ jobs:
     - name: 📥 Checkout repository
       uses: actions/checkout@v1
 
+    - name: 🧰 Select optimal cabal version
+      run: |
+        case "$OS" in
+          Windows_NT)   echo "CABAL_VERSION=3.4.0.0-rc5"  >> $GITHUB_ENV;;
+          *)            echo "CABAL_VERSION=3.4.0.0-rc4"  >> $GITHUB_ENV;;
+        esac
+
     - name: 🧰 Setup Haskell
-      uses: actions/setup-haskell@v1
+      uses: haskell/actions/setup@v1
       id: setup-haskell
       with:
         ghc-version: ${{ matrix.ghc }}
-        cabal-version: '3.2.0.0'
-
-    - name: 🩹 Patch GHC 8.10.2 linker
-      if: matrix.os == 'windows-latest' && matrix.ghc == '8.10.2'
-      run: |
-        sed -i \
-          's|C:/GitLabRunner/builds/2WeHDSFP/0/ghc/ghc/inplace/mingw/bin/ld.exe|C:/ProgramData/chocolatey/lib/ghc.8.10.2/tools/ghc-8.10.2/mingw/bin/ld.exe|g' \
-          C:/ProgramData/chocolatey/lib/ghc.8.10.2/tools/ghc-8.10.2/lib/settings
+        cabal-version: ${{ env.CABAL_VERSION }}
 
     - name: 🔄 Cabal update
       run: cabal update
 
     - name: 🛠 Cabal Configure
       run: cabal configure --builddir="$CABAL_BUILDDIR" --enable-tests --enable-benchmarks --write-ghc-environment-files=always
 
+    - name: 💾 Record dependencies
+      run: |
+        cat ${{ env.PLAN_JSON }} | jq -r '."install-plan"[].id' | sort | uniq > dependencies.txt
+
+    - name: 💾 Set cache version
+      run: echo "CACHE_VERSION=9w76Z3Q" >> $GITHUB_ENV
+
     - name: 💾 Cache Dependencies
-      id: cache
       uses: actions/cache@v2
       with:
         path: ${{ steps.setup-haskell.outputs.cabal-store }}
-        key: cache-cabal-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('bech32/bech32.cabal') }}-${{ hashFiles('bech32-th/bech32-th.cabal') }}
+        key: cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+        restore-keys: |
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-
```

Comment:
Why would we want that latter cache restoration option? If the dependencies have changed and there's no cache hit we ought to rebuild, don't we?

Reply:
The second cache key is a fallback.

Github Actions will use the cache key in an exact match, but if it does not exist, it will fallback to the most recent cache key (in git history) with the given prefix. See https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows

This is useful for example when someone tweaks bounds on one of the dependencies.  Without the fallback, this would trigger a full rebuild, whereas with the fallback, this will only rebuild affected dependencies (at the small cost of a slightly larger cache).

Something like this is more important for larger projects that take an hour or more to build, but it seems a useful thing to have generally.

If you need to force a full rebuild (for example the cache is somehow corrupted), you can achieve that by redefining `CACHE_VERSION` to another random string on line `57`.

There is a better way to cache using https://github.com/haskell-works/cabal-cache, that is always optimal, but that requires setting up an S3 bucket and Github credentials for the S3 bucket, so I've avoided that for now and stayed with a simple cache with reasonable trade offs.

Reply:
Ok fair enough.